Source/Core/VideoBackends/D3D12/D3DQueuedCommandList.cpp

```diff
@@ -0,0 +1,1217 @@
+
+					item += sizeof(PresentArguments) + sizeof(D3DQueueItemType) * 2;
+					break;
+				}
+
+				case D3DQueueItemType::ResetCommandList:
+				{
+					CheckHR(command_list->Reset(reinterpret_cast<D3DQueueItem*>(item)->ResetCommandList.allocator, nullptr));
+
+					item += sizeof(ResetCommandListArguments) + sizeof(D3DQueueItemType) * 2;
+					break;
+				}
+
+				case D3DQueueItemType::ResetCommandAllocator:
+				{
+					CheckHR(reinterpret_cast<D3DQueueItem*>(item)->ResetCommandAllocator.allocator->Reset());
+
+					item += sizeof(ResetCommandAllocatorArguments) + sizeof(D3DQueueItemType) * 2;
+					break;
+				}
+
+				case D3DQueueItemType::FenceGpuSignal:
+				{
+					CheckHR(parent_queued_command_list->m_command_queue->Signal(reinterpret_cast<D3DQueueItem*>(item)->FenceGpuSignal.fence, reinterpret_cast<D3DQueueItem*>(item)->FenceGpuSignal.fence_value));
+
+					item += sizeof(FenceGpuSignalArguments) + sizeof(D3DQueueItemType) * 2;
+					break;
+				}
+
+				case D3DQueueItemType::FenceCpuSignal:
+				{
+					CheckHR(reinterpret_cast<D3DQueueItem*>(item)->FenceCpuSignal.fence->Signal(reinterpret_cast<D3DQueueItem*>(item)->FenceCpuSignal.fence_value));
+
+					item += sizeof(FenceCpuSignalArguments) + sizeof(D3DQueueItemType) * 2;
+					break;
+				}
+
+				case D3DQueueItemType::Stop:
+
+					// Use a goto to break out of the loop, since we can't exit the loop from
+					// within a switch statement. We could use a separate 'if' after the switch,
+					// but that was the highest source of overhead in the function after profiling.
+					// http://stackoverflow.com/questions/1420029/how-to-break-out-of-a-loop-from-inside-a-switch
+
+					item += sizeof(D3DQueueItemType) * 2;
+
+					if (item - queue_array > s_queue_array_size / 3)
+					{
+						item = queue_array;
+					}
+
+					goto exitLoop;
+			}
+		}
+
+		exitLoop:
+
+		parent_queued_command_list->m_queue_array_front = static_cast<UINT>(item - queue_array);
+	}
+}
+
+ID3D12QueuedCommandList::ID3D12QueuedCommandList(ID3D12GraphicsCommandList* backing_command_list, ID3D12CommandQueue* backing_command_queue) :
+	m_command_list(backing_command_list),
+	m_command_queue(backing_command_queue)
+{
+	memset(m_queue_array, 0, sizeof(m_queue_array));
+
+	m_queue_array_back = m_queue_array;
+
+	m_begin_execution_event = CreateSemaphore(nullptr, 0, 256, nullptr);
+
+	m_background_thread = CreateThread(nullptr, 0, BackgroundThreadFunction, this, 0, &m_background_thread_id);
+}
+
+ID3D12QueuedCommandList::~ID3D12QueuedCommandList()
+{
+	TerminateThread(m_background_thread, 0);
+	CloseHandle(m_background_thread);
+
+	CloseHandle(m_begin_execution_event);
+}
+
+void ID3D12QueuedCommandList::QueueExecute()
+{
+	reinterpret_cast<D3DQueueItem*>(m_queue_array_back)->Type = D3DQueueItemType::ExecuteCommandList;
+
+	m_queue_array_back += sizeof(ExecuteCommandListArguments) + sizeof(D3DQueueItemType) * 2;
```

Comment:
I know I've mentioned this before, but the lack of bounds check for these methods is concerning, especially if the worker thread got stalled for some reason.